### PR TITLE
fix: Fixes project flaky tests

### PR DESCRIPTION
**Pull request by lantoli**
## Description

Jira ticket: [INTMDB-1341](https://jira.mongodb.org/browse/INTMDB-1341)

Fixes project flaky tests

## Type of change:

- [X] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [X] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [X] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [X] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [X] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [X] I have added any necessary documentation (if appropriate)
- [X] I have run make fmt and formatted my code
- [X] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.

### CODE REVIEW

File: .github/workflows/acceptance-tests.yml [hunk 1]
```diff
@@ -335,7 +335,6 @@ jobs:
           MONGODB_ATLAS_ORG_ID: ${{ vars.MONGODB_ATLAS_ORG_ID_CLOUD_DEV }}
           MONGODB_ATLAS_BASE_URL: ${{ vars.MONGODB_ATLAS_BASE_URL }}
           MONGODB_ATLAS_PROJECT_OWNER_ID: ${{ vars.MONGODB_ATLAS_PROJECT_OWNER_ID }}
-          MONGODB_ATLAS_API_KEYS_IDS: ${{ vars.MONGODB_ATLAS_API_KEYS_IDS }}
```
Comment (lantoli) [inline .github/workflows/acceptance-tests.yml:338]: not used

File: internal/testutil/acc/pre_check.go [hunk 2]
```diff
@@ -45,10 +46,26 @@ func PreCheckAtlasUsername(tb testing.TB) {
 	}
 }
 
-func PreCheckTeamsIds(tb testing.TB) {
-	if os.Getenv("MONGODB_ATLAS_TEAMS_IDS") == "" {
-		tb.Skip("`MONGODB_ATLAS_TEAMS_IDS` must be set for Projects acceptance testing")
+func PreCheckProjectTeamsIds(tb testing.TB, min int) {
```
Comment (lantoli) [inline internal/testutil/acc/pre_check.go]: improve teamsIds checks, and fail tests if pre-check is not met

File: internal/service/project/data_source_projects.go [hunk 3]
```diff
@@ -171,17 +171,14 @@ func (d *ProjectsDS) Read(ctx context.Context, req datasource.ReadRequest, resp
 }
 
 func populateProjectsDataSourceModel(ctx context.Context, conn *matlas.Client, connV2 *admin.APIClient, stateModel *tfProjectsDSModel, projectsRes *matlas.Projects) error {
```
Comment (lantoli) [inline internal/service/project/data_source_projects.go:173]: the fix is in this func

File: internal/service/project/data_source_project_test.go [hunk 4]
```diff
@@ -15,26 +14,23 @@ import (
 )
 
 func TestAccProjectDSProject_byID(t *testing.T) {
-	projectName := acctest.RandomWithPrefix("test-acc")
-	orgID := os.Getenv("MONGODB_ATLAS_ORG_ID")
-	teamsIds := strings.Split(os.Getenv("MONGODB_ATLAS_TEAMS_IDS"), ",")
-	if len(teamsIds) < 2 {
-		t.Skip("`MONGODB_ATLAS_TEAMS_IDS` must have 2 team ids for this acceptance testing")
```
Comment (lantoli) [inline internal/service/project/data_source_project_test.go:22]: improve pre-checks and fail if they're not met instead of just skipping the test

File: internal/service/project/data_source_projects.go [hunk 5]
```diff
@@ -171,17 +171,14 @@ func (d *ProjectsDS) Read(ctx context.Context, req datasource.ReadRequest, resp
...
-	results := make([]*tfProjectDSModel, len(projectsRes.Results))
-
-	for i, project := range projectsRes.Results {
+	results := make([]*tfProjectDSModel, 0, len(projectsRes.Results))
+	for _, project := range projectsRes.Results {
 		atlasTeams, atlasLimits, atlasProjectSettings, err := getProjectPropsFromAPI(ctx, conn, connV2, project.ID)
-		if err != nil {
-			return fmt.Errorf("error while getting project properties for project %s: %v", project.ID, err.Error())
+		if err == nil { // if the project is still valid, e.g. could have just been deleted
```
Comment (marcosuma) [inline internal/service/project/data_source_projects.go:177]: should we fail if `err != nil` ?

File: internal/testutil/acc/pre_check.go [hunk 6]
```diff
@@ -45,10 +46,26 @@ func PreCheckAtlasUsername(tb testing.TB) {
...
+		return
 	}
+	teamsIds := strings.Split(envVar, ",")
+	if count := len(teamsIds); count < min {
+		tb.Fatalf("`MONGODB_ATLAS_TEAMS_IDS` must have at least %d team ids for this acceptance testing, has %d", min, count)
+	}
+}
+
+func GetProjectTeamsIds(pos int) string {
+	envVar := os.Getenv("MONGODB_ATLAS_TEAMS_IDS")
```
Comment (marcosuma) [inline internal/testutil/acc/pre_check.go:62]: nit: can you please pass the name of the environment variable as input of the function?

File: internal/testutil/acc/pre_check.go (hunk 2 above)
Comment (marcosuma) [inline internal/testutil/acc/pre_check.go]: nit: can you please pass the name of the environment variable as input of the function?

File: internal/service/project/data_source_project_test.go [hunk 7]
```diff
@@ -15,26 +14,23 @@ import (
...
-	}
-
+	var (
+		projectName = acctest.RandomWithPrefix("test-acc")
+		orgID       = os.Getenv("MONGODB_ATLAS_ORG_ID")
+	)
 	resource.ParallelTest(t, resource.TestCase{
-		PreCheck:                 func() { acc.PreCheckBasic(t); acc.PreCheckTeamsIds(t) },
+		PreCheck:                 func() { acc.PreCheckBasic(t); acc.PreCheckProjectTeamsIds(t, 2) },
```
Comment (marcosuma) [inline internal/service/project/data_source_project_test.go]:
```suggestion
		PreCheck:                 func() { acc.PreCheckBasic(t); acc.PreCheckProjectTeamsIdsWithMinSize(t, 2) },
```

nit: for clarity I would suggest a more specific method name

File: internal/service/project/data_source_projects.go (hunk 5 above)
Comment (lantoli) [reply internal/service/project/data_source_projects.go:177]: no, that was the problem before.  Imagine we get the list of projects for the org, but then some of them are deleted.  It's not really a failure, so we just ignore it if it's deleted after it was in the list.   This was the reason for project flaky test.

File: internal/testutil/acc/pre_check.go (hunk 2 above)
Comment (marcosuma) [inline internal/testutil/acc/pre_check.go]:
```suggestion
func PreCheckProjectTeamsIds(tb testing.TB, minProjectCount int) {
```

**Review by marcosuma — APPROVED** [review]
few minor comments, LGTM overall

File: internal/testutil/acc/pre_check.go (hunk 6 above)
Comment (lantoli) [reply internal/testutil/acc/pre_check.go:62]: I think it's better in this way as PreCheckProjectTeamsIds and GetProjectTeamsIds work together for teamsIds and they both use the same env.var

File: internal/testutil/acc/pre_check.go (hunk 2 above)
Comment (lantoli) [reply internal/testutil/acc/pre_check.go]: same, I think it's better to have the env.var in both methods as they're used only for teamsIds, these methods are not intended to be generic for other purporses.

File: internal/service/project/data_source_project_test.go (hunk 7 above)
Comment (lantoli) [reply internal/service/project/data_source_project_test.go]: I can do that but I normally like to say "with..." when there are multiple related methods, there is only one method in this case

File: internal/testutil/acc/pre_check.go (hunk 2 above)
Comment (lantoli) [reply internal/testutil/acc/pre_check.go]: I can change it, although I tried to follow "go idiomatic" in the sense the I think "min" in this context is clear, but I agree I can make it more specific here.

Comment (lantoli) [reply internal/testutil/acc/pre_check.go]: renamed to minTeamsCount (although it looks a bit redudant to me as Teams is already in function name)

File: internal/service/project/data_source_project_test.go (hunk 7 above)
Comment (lantoli) [reply internal/service/project/data_source_project_test.go]: renamed to: func PreCheckProjectTeamsIdsWithMinCount(tb testing.TB, minTeamsCount int)

File: internal/service/project/data_source_projects.go (hunk 5 above)
Comment (marcosuma) [reply internal/service/project/data_source_projects.go:177]: ok noted, but then I guess we're ok to ignore 5xx errors?
What I am trying to understand here is: if there is a 5xx error and we ignore that and we move on, will we get a different test behaviour that might hide some problem?

File: internal/testutil/acc/pre_check.go (hunk 6 above)
Comment (marcosuma) [reply internal/testutil/acc/pre_check.go:62]: > work together 

I think this is an assumption that might not be always true, unless you call the `precheck...` in the `get...` method right?

File: internal/service/project/data_source_projects.go (hunk 5 above)
Comment (lantoli) [reply internal/service/project/data_source_projects.go:177]: I've got 500 sometimes when doing something in a project that is being closed, so we could get also a 500 and it doesn't mean a real error but that the project is closing.   So I think it's hard in this case to differentiate a "real" error from an error caused by the project being closed. Also chances are that if the project is in the list and it's not closed we'll get the details without any problem

File: internal/testutil/acc/pre_check.go (hunk 6 above)
Comment (lantoli) [reply internal/testutil/acc/pre_check.go:62]: if you don't call precheck you won't check that you have the min teams you need but it will still work (returns "" when you don't have that pos).  
but if you don't call the precheck method, i don't see how passing the env.var can help there.

(They have the same name part ProjectTeamsIds and are located together in the file to make it easier to see that they are intended to work together)

File: internal/service/project/data_source_projects.go (hunk 5 above)
Comment (AgustinBettati) [reply internal/service/project/data_source_projects.go:177]: Understand this an edge case that is likely only appearing  in our acceptance tests execution: fetching and creating/deleting project in parallel. Curious if a potential fix to this problem could be defining the data source test to run sequentially (`resource.Test(` over `resource.ParallelTest(`) to reduce the chances of this scenario. (just to keep in mind for other flaky tests that we might have)

Comment (lantoli) [reply internal/service/project/data_source_projects.go:177]: that would probably fix the tests, also to use the concurrency facility in GH actions so project acc and migration tests don't run in parallel.  But at the end I thought this is a realistic use case that can happen sometimes and we don't want the DS to fail if that happens, that's why I focused on improving the prod. code.
